File: jz/jz_pipeline/TriangleTree.cs
using Microsoft.Xna.Framework;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Runtime.InteropServices;

namespace jz
{
    public class TriangleTree
    {
        #region Protected members
        [StructLayout(LayoutKind.Explicit)]
        protected struct Node
        {
            public static Node Create()
            {
                Node ret = default(Node);
                ret.Sibling = 0u;
                ret.mFlags = 0u;

                return ret;
            }

            const UInt32 kAxisMask = (1 << 0) | (1 << 1);
            const UInt32 kFbMask = (1 << 2);
            const UInt32 kLeafMask = (1 << 3);

            public Axis Axis
            {
                get
                {
                    Axis ret = (Axis)(mFlags & kAxisMask);
                    return ret;
                }

                set
                {
                    mFlags &= ~kAxisMask;
                    mFlags |= (((UInt32)value) & kAxisMask);
                }
            }

            public float SplitPos
            {
                get
                {
                    return mSplitPos;
                }

                set
                {
                    UInt32 flags = (mFlags & (kAxisMask | kFbMask | kLeafMask));
                    mSplitPos = value;
                    mFlags &= ~(kAxisMask | kFbMask | kLeafMask);
                    mFlags |= flags;
                }
            }

            public bool IsLeaf { get { return ((mFlags & kLeafMask) != 0); } }
            public bool IsBack { get { return ((mFlags & kFbMask) == 0); } }
            public bool IsFront { get { return ((mFlags & kFbMask) != 0); } }

            public void SetBack() { mFlags &= ~kFbMask; }
            public void SetFront() { mFlags |= kFbMask; }
            public void SetLeaf() { mFlags |= kLeafMask; }

            public void Write(BinaryWriter w)
            {
                w.Write(mFlags);
                w.Write(Sibling);
            }

            [FieldOffset(0)]
            private float mSplitPos;
            [FieldOffset(0)]
            private UInt32 mFlags;

            [FieldOffset(4)]
            public UInt32 Sibling;
            [FieldOffset(4)]
            public UInt32 TriangleIndex;
        };

        protected List<Node> mNodes = new List<Node>();
        protected List<UInt16> mTriangles = new List<UInt16>();
        protected List<Vector3> mVertices = new List<Vector3>();
        #endregion

        #region Private members
        private BoundingBox mTotalAABB = Utilities.kZeroBox;
        private Vector3 CalculateMean(List<UInt32> e)
        {
            Vector3 ret = Vector3.Zero;
            int count = e.Count;

            if (count > 0)
            {
                for (int i = 0; i < count; i++)
                {
                    Triangle t = GetTriangle(e[i]);
                    ret += t.Center;
                }
                ret *= (1.0f / (float)count);
            }

            return ret;
        }

        private Vector3 CalculateVariance(Vector3 mean, List<UInt32> e)
        {
            Vector3 ret = Vector3.Zero;

            int count = e.Count;
            if (count > 1)
            {
                for (int i = 0; i < count; i++)
                {
                    Triangle t = GetTriangle(e[i]);

                    Vector3 a = (t.Center - mean);
                    ret += (a * a);
                }

                ret = (ret / ((float)(count - 1)));
            }

            return ret;
        }

        private BoundingBox CalculateTotalAABB(List<UInt32> e)
        {
            Vector3 min = Utilities.kMaxVector3;
            Vector3 max = Utilities.kMinVector3;

            int size = e.Count;
            for (int i = 0; i < size; i++)
            {
                Triangle t = GetTriangle(e[i]);

                min = Vector3.Min(min, Vector3.Min(t.P0, Vector3.Min(t.P1, t.P2)));
                max = Vector3.Max(max, Vector3.Max(t.P0, Vector3.Max(t.P1, t.P2)));
            }

            return new BoundingBox(min, max);
        }

        private Triangle GetTriangle(UInt32 i)
        {
            UInt32 kIndex = (i * 3u);

            Triangle ret = new Triangle(
                mVertices[(int)mTriangles[(int)(kIndex + 0u)]],
                mVertices[(int)mTriangles[(int)(kIndex + 1u)]],
                mVertices[(int)mTriangles[(int)(kIndex + 2u)]]);

            return ret;
        }

        private void Split(float pos, Axis axis,
            List<UInt32> e, List<UInt32> front, List<UInt32> back,
            out float frontPos, out float backPos)
        {
            frontPos = pos;
            backPos = pos;

            int size = e.Count;
            for (int i = 0; i < size; i++)
            {
                Triangle t = GetTriangle(e[i]);
                float center = Utilities.GetElement(t.Center, (int)axis);

                if (pos <= center)
                {
                    frontPos = Utilities.Min(frontPos, Utilities.GetElement(t.Min, (int)axis));
                    front.Add(e[i]);
                    e[i] = e[size - 1];
                    size--; i--;
                }
                else if (pos > center)
                {
                    backPos = Utilities.Max(backPos, Utilities.GetElement(t.Max, (int)axis));
                    back.Add(e[i]);
                    e[i] = e[size - 1];
                    size--; i--;
                }
            }

            Debug.Assert(front.Count + back.Count == e.Count);
            Debug.Assert(size == 0);
            e.Clear();
        }

        private void Split(int aIndex, BoundingBox aParentAABB, List<UInt32> e)
        {
            Debug.Assert(e.Count > 0);

            if (e.Count == 1)
            {
                Node n = mNodes[aIndex];
                n.SetLeaf();
                n.TriangleIndex = e[0];
                e.Clear();
                mNodes[aIndex] = n;
                return;
            }

            Vector3 mean = CalculateMean(e);
            Vector3 variance = CalculateVariance(mean, e);
            float max = Utilities.Max(variance.X, variance.Y, variance.Z);

            Axis axis = (max == variance.Z) ? Axis.Z : ((max == variance.X) ? Axis.X : Axis.Y);
            float pos = Utilities.GetElement(Utilities.GetCenter(aParentAABB), (int)axis);

            List<UInt32> front = new List<UInt32>();
            List<UInt32> back = new List<UInt32>();
            float frontPos = 0.0f;
            float backPos = 0.0f;
            Split(pos, axis, e, front, back, out frontPos, out backPos);
            Debug.Assert(e.Count == 0);

            // It's possible that splitting at the box center places all the objects on one side of the split.
            if ((front.Count == 0 && back.Count > 1) || (back.Count == 0 && front.Count > 1))
            {
                e.AddRange(front); front.Clear();
                e.AddRange(back); back.Clear();

                pos = Utilities.GetElement(mean, (int)axis);
                Split(pos, axis, e, front, back, out frontPos, out backPos);

                Debug.Assert(!((front.Count == 0 && back.Count > 1) || (back.Count == 0 && front.Count > 1)));
            }

            #region Create front node
            if (front.Count > 0)
            {
                mNodes.Add(Node.Create());
                Node n = mNodes[mNodes.Count - 1];
                n.Axis = axis;
                n.SplitPos = frontPos;
                n.SetFront();
                mNodes[mNodes.Count - 1] = n;

                BoundingBox frontAABB = aParentAABB;
                Utilities.SetElement(ref frontAABB.Min, (int)axis, frontPos);

                Split(mNodes.Count - 1, frontAABB, front);
            }
            #endregion

            #region Create back node
            if (back.Count > 0)
            {
                mNodes.Add(Node.Create());
                Node n = mNodes[mNodes.Count - 1];
                n.Axis = axis;
                n.SplitPos = backPos;
                n.SetBack();
                mNodes[mNodes.Count - 1] = n;

                BoundingBox backAABB = aParentAABB;
                Utilities.SetElement(ref backAABB.Max, (int)axis, backPos);

                Split(mNodes.Count - 1, backAABB, back);
            }
            #endregion

            #region Set node sibling reference
            {
                Node n = mNodes[aIndex];
                n.Sibling = (UInt32)mNodes.Count;
                mNodes[aIndex] = n;
            }
            #endregion
        }
        #endregion

        #region Public members
        public void Build(List<UInt16> aTriangles, List<Vector3> aVertices)
        {
            mTriangles.Clear();
            mTriangles.AddRange(aTriangles);
            mVertices.Clear();
            mVertices.AddRange(aVertices);

            List<UInt32> t = new List<UInt32>();
            UInt32 count = (UInt32)mTriangles.Count;
            for (UInt32 i = 0; i < count; i += 3) { t.Add(i / 3u); }

            mTotalAABB = CalculateTotalAABB(t);

            mNodes.Clear();

            if (t.Count > 0)
            {
                #region Insert a root node
                mNodes.Add(Node.Create());
                Node n = mNodes[0];
                n.Axis = Axis.Z;
                n.SetFront();
                n.SplitPos = mTotalAABB.Min.Z;
                mNodes[0] = n;
                #endregion
                Split(0, mTotalAABB, t);
                if (mNodes.Count == 1) { mNodes[0].SetLeaf(); }
            }
        }

        public void Write(BinaryWriter w)
        {
            w.Write((UInt32)mTriangles.Count);
            foreach (UInt16 e in mTriangles) { w.Write(e); }

            w.Write((UInt32)mVertices.Count);
            foreach (Vector3 e in mVertices) { w.Write(e.X); w.Write(e.Y); w.Write(e.Z); }

            w.Write((UInt32)mNodes.Count);
            foreach (Node e in mNodes) { e.Write(w); }

            w.Write(mTotalAABB.Min.X); w.Write(mTotalAABB.Min.Y); w.Write(mTotalAABB.Min.Z);
            w.Write(mTotalAABB.Max.X); w.Write(mTotalAABB.Max.Y); w.Write(mTotalAABB.Max.Z);
        }
        #endregion
    }
     
}

File: jz/jz_system/TriangleTree.cpp
//
// Copyright (c) 2009 Joseph A. Zupko
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.
// 

#include <jz_system/Files.h>
#include <jz_system/ReadHelpers.h>
#include <jz_system/TriangleTree.h>
#include <algorithm>

namespace jz
{
    namespace system
    {

        void TriangleTree::Node::Read(IReadFilePtr& p)
        {
            mFlags = ReadUInt32(p);
            Sibling = ReadUInt32(p);
        }

        Vector3 TriangleTree::CalculateMean(const vector<u32>& e)
        {
            Vector3 ret = Vector3::kZero;
            size_t size = e.size();

            if (size > 0u)
            {
                for (size_t i = 0u; i < size; i++)
                {
                    Triangle3D t = GetTriangle(e[i]);
                    ret += t.GetCenter();
                }
                ret *= (1.0f / (float)size);
            }

            return ret;
        }

        Vector3 TriangleTree::CalculateVariance(const Vector3& mean, const vector<u32>& e)
        {
            Vector3 ret = Vector3::kZero;

            size_t size = e.size();
            if (size > 1)
            {
                for (size_t i = 0; i < size; i++)
                {
                    Triangle3D t = GetTriangle(e[i]);

                    Vector3 a = (t.GetCenter() - mean);
                    ret += (a * a);
                }

                ret = (ret / ((float)(size - 1u)));
            }

            return ret;
        }

        BoundingBox TriangleTree::CalculateTotalAABB(const vector<u32>& aTriangles)
        {
            Vector3 min = Vector3::kMax;
            Vector3 max = Vector3::kMin;

            size_t size = aTriangles.size();
            for (size_t i = 0u; i < size; i++)
            {
                Triangle3D t = GetTriangle(aTriangles[i]);

                min = Vector3::Min(min, Vector3::Min(t.P0, Vector3::Min(t.P1, t.P2)));
                max = Vector3::Max(max, Vector3::Max(t.P0, Vector3::Max(t.P1, t.P2)));
            }

            return BoundingBox(min, max);
        }

        Triangle3D TriangleTree::GetTriangle(u32 i)
        {
            const u32 kIndex = (i * 3u);

            JZ_ASSERT(kIndex < mTriangles.size());

            Triangle3D ret = Triangle3D(
                mVertices[mTriangles[(kIndex + 0u)]],
                mVertices[mTriangles[(kIndex + 1u)]],
                mVertices[mTriangles[(kIndex + 2u)]]);

            return ret;
        }

        void TriangleTree::Split(float pos, Axis::Id axis,
            vector<u32>& e, vector<u32>& front, vector<u32>& back,
            float& frontPos, float& backPos)
        {
            frontPos = pos;
            backPos = pos;

            size_t size = e.size();
            for (size_t i = 0u; i < size; i++)
            {
                Triangle3D t = GetTriangle(e[i]);
                float center = t.GetCenter()[axis];

                if (pos <= center)
                {
                    frontPos = jz::Min(frontPos, t.GetMin()[axis]);
                    front.push_back(e[i]);
                    e[i] = e[size - 1u];
                    size--; i--;
                }
                else if (pos > center)
                {
                    backPos = jz::Max(backPos, t.GetMax()[axis]);
                    back.push_back(e[i]);
                    e[i] = e[size - 1u];
                    size--; i--;
                }
            }

            JZ_ASSERT(front.size() + back.size() == e.size());
            JZ_ASSERT(e.size() == 0u);
            e.clear();
        }

        void TriangleTree::Split(size_t aIndex, const BoundingBox& aParentAABB, vector<u32>& e)
        {
            JZ_ASSERT(e.size() > 0u);

            if (e.size() == 1u)
            {
                mNodes[aIndex].SetLeaf();
                mNodes[aIndex].TriangleIndex = e[0];
                e.clear();
                return;
            }

            Vector3 mean = CalculateMean(e);
            Vector3 variance = CalculateVariance(mean, e);
            float max = jz::Max(variance.X, variance.Y, variance.Z);

            Axis::Id axis = (max == variance.Z) ? Axis::kZ : ((max == variance.X) ? Axis::kX : Axis::kY);
            float pos = aParentAABB.Center()[axis];

            vector<u32> front;
            vector<u32> back;
            float frontPos = 0.0f;
            float backPos = 0.0f;
            Split(pos, axis, e, front, back, frontPos, backPos);
            JZ_ASSERT(e.size() == 0u);

            // It's possible that splitting at the box center places all the objects on one side of the split.
            if ((front.size() == 0u && back.size() > 1u) || (back.size() == 0u && front.size() > 1u))
            {
                copy(front.begin(), front.end(), e.end());
                front.clear();
                copy(back.begin(), back.end(), e.end());
                back.clear();

                pos = mean[axis];
                Split(pos, axis, e, front, back, frontPos, backPos);
                JZ_ASSERT(!((front.size() == 0u && back.size() > 1u) || (back.size() == 0u && front.size() > 1u)));
            }

            #pragma region Create front node
            if (front.size() > 0u)
            {
                mNodes.push_back(Node());
                mNodes[mNodes.size() - 1u].SetAxis(axis);
                mNodes[mNodes.size() - 1u].SetSplitPos(frontPos);
                mNodes[mNodes.size() - 1u].SetFront();

                BoundingBox frontAABB = aParentAABB;
                frontAABB.Min[axis] = frontPos;

                Split(mNodes.size() - 1u, frontAABB, front);
            }
            #pragma endregion

            #pragma region Create back node
            if (back.size() > 0u)
            {
                mNodes.push_back(Node());
                mNodes[mNodes.size() - 1u].SetAxis(axis);
                mNodes[mNodes.size() - 1u].SetSplitPos(backPos);
                mNodes[mNodes.size() - 1u].SetBack();

                BoundingBox backAABB = aParentAABB;
                backAABB.Max[axis] = backPos;

                Split(mNodes.size() - 1u, backAABB, back);
            }
            #pragma endregion

            mNodes[aIndex].Sibling = mNodes.size();
        }

        void TriangleTree::Build(const MemoryBuffer<u16>& aTriangles, const MemoryBuffer<Vector3>& aVertices)
        {
            mTriangles.clear();
            mTriangles = aTriangles;
            mVertices.clear();
            mVertices = aVertices;

            vector<u32> t;
            size_t size = mTriangles.size();
            for (size_t i = 0u; i < size; i += 3) { t.push_back(i / 3); }

            mTotalAABB = CalculateTotalAABB(t);

            mNodes.clear();
            #pragma region Insert a root node
            mNodes.resize(1);
            mNodes[0].SetAxis(Axis::kZ);
            mNodes[0].SetFront();
            mNodes[0].SetSplitPos(mTotalAABB.Min.Z);
            #pragma endregion        
            Split(0u, mTotalAABB, t);
        }

        void TriangleTree::Read(IReadFilePtr& pFile)
        {
            mTriangles.clear();
            mVertices.clear();
            mNodes.clear();

            ReadBuffer(pFile, mTriangles);
            ReadBuffer(pFile, mVertices);

            size_t size = ReadSizeT(pFile);
            for (size_t i = 0u; i < size; i++)
            {
                Node node;
                node.Read(pFile);

                mNodes.push_back(node);
            }

            mTotalAABB = ReadBoundingBox(pFile);
        }

    }
}
